### PR TITLE
Moto server

### PR DESCRIPTION
**Pull request by martindurant**
Changes moto for s3 tests from monkeypatched/mocking to server mode. This allows aiobotocore exceptions to raise correctly, needed for tests to pass; the change is required by the upcoming async release of s3fs, but also works for old sync (botocore) version. A release of s3fs without this change would break pandas tests.

Note: since I now need to pass storage_options to the various s3 calls in the tests to make them pass (giving the endpoint of the moto s3 server), I needed to also plumb storage_options through the excel IO, which was previously missing, and update the excel tests.

- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by martindurant** [comment]
Good to go.
Does something like this need a whatsnew, when it only affects the tests and won't be seen by users?
I could add whatsnew about the excel integration, although it's pretty niche.

Sorry about the long commit list, it contains the original list of commits from #35381 (i.e., wasn't squashed here, only in master).

File: pandas/io/excel/_odfreader.py [hunk 1]
```diff
@@ -18,11 +18,13 @@ class _ODFReader(_BaseExcelReader):
     ----------
     filepath_or_buffer: string, path to be parsed or
         an open readable stream.
+    storage_options: dict (optional)
```
Comment (jreback) [inline pandas/io/excel/_odfreader.py:21]: need a versionadded tag

File: pandas/io/excel/_openpyxl.py [hunk 2]
```diff
@@ -467,17 +467,21 @@ def write_cells(
 
 
 class _OpenpyxlReader(_BaseExcelReader):
-    def __init__(self, filepath_or_buffer: FilePathOrBuffer) -> None:
+    def __init__(
+        self, filepath_or_buffer: FilePathOrBuffer, storage_options=None
```
Comment (jreback) [inline pandas/io/excel/_openpyxl.py:471]: can you type wiht Optional[StorageOptions]

File: pandas/io/excel/_openpyxl.py [hunk 3]
```diff
@@ -467,17 +467,21 @@ def write_cells(
...
+    ) -> None:
         """
         Reader using openpyxl engine.
 
         Parameters
         ----------
         filepath_or_buffer : string, path object or Workbook
             Object to be parsed.
+        storage_options: dict (optional)
```
Comment (jreback) [inline pandas/io/excel/_openpyxl.py:480]: same on all of these

File: pandas/tests/io/conftest.py [hunk 4]
```diff
@@ -72,12 +76,30 @@ def add_tips_files(bucket_name):
                     conn.Bucket(bucket_name).put_object(Key=s3_key, Body=f)
 
         try:
-            s3 = moto.mock_s3()
-            s3.start()
+            import shlex
```
Comment (jreback) [inline pandas/tests/io/conftest.py:79]: can you add some comments on what you are doing

File: pandas/tests/io/parser/test_network.py [hunk 5]
```diff
@@ -52,6 +52,9 @@ def tips_df(datapath):
     return read_csv(datapath("io", "data", "csv", "tips.csv"))
 
 
+s3so = dict(client_kwargs={"endpoint_url": "http://127.0.0.1:5555/"})
```
Comment (jreback) [inline pandas/tests/io/parser/test_network.py:55]: can we make these a fixture so don't have to define everywhere

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/excel/_odfreader.py (hunk 1 above)
Comment (martindurant) [reply pandas/io/excel/_odfreader.py:21]: Put into `pandas.io.excel._base._read_excel_doc`, so shows up for `pd.read_excel` (i.e., users are not expected to access these inner modules/classes)

File: pandas/io/excel/_openpyxl.py (hunk 2 above)
Comment (martindurant) [reply pandas/io/excel/_openpyxl.py:471]: OK

File: pandas/io/excel/_openpyxl.py (hunk 3 above)
Comment (martindurant) [reply pandas/io/excel/_openpyxl.py:480]: Typed using StorageOptions

File: pandas/tests/io/parser/test_network.py (hunk 5 above)
Comment (martindurant) [reply pandas/tests/io/parser/test_network.py]: done

File: pandas/tests/io/conftest.py (hunk 4 above)
Comment (martindurant) [reply pandas/tests/io/conftest.py:87]: done

**Comment by martindurant** [comment]
@jreback : went green again

File: pandas/tests/io/conftest.py (hunk 4 above)
Comment (jreback) [reply pandas/tests/io/conftest.py:87]: you can move the imports to the top of the file that can (subprocess, time shlex)

import requests can be outside the try

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, comment on the impl & can you add this issue number to the whatsnew note where you added storage options; i think maybe need to update as now excel works

**Comment by martindurant** [comment]
I didn't understand "comment on the impl" - but I did move the imports and add a sentence to whatsnew for excel.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @TomAugspurger if any comments.

**Comment by TomAugspurger** [comment]
I see one environment where there's moto but not flask, which results in skips like https://travis-ci.org/github/pandas-dev/pandas/jobs/717300240#L4633.

- ci/deps/azure-37-locale.yaml

Also, at least some of the tests are being skipped for other builds.

- https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=40289&view=logs&j=4b4d380c-92b5-58cf-5b80-2c6c9b1d5805&t=e685736b-3705-5348-f648-7de631559230&l=2417 (moto too old)
- azure-38-locale: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=40289&view=logs&j=b1c7b65e-b3ce-541a-7fd5-29b4ba56ce18&t=f832e1c9-f124-52e9-664c-6040860879fb&l=2159 (no s3fs)
- azure-37-locale: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=40289&view=logs&j=a3a13ea8-7cf0-5bdb-71bb-6ac8830ae35c&t=99777837-3d45-5d3a-4778-aac6f4679b04&l=2217 (no s3fs)
- travis-37-cov: https://travis-ci.org/github/pandas-dev/pandas/jobs/717300242#L4113 (moto too old)

I think we should take the files that have s3fs and ensure that they have new enough versions of moto and s3fs. That would be

* ci/deps/azure-37-slow.yaml
* ci/deps/azure-windows-37.yaml
* ci/deps/travis-37-cov.yaml
* ci/deps/travis-37.yaml

File: pandas/io/excel/_base.py [hunk 6]
```diff
@@ -837,14 +850,16 @@ class ExcelFile:
...
+    _engines: Mapping[str, Any] = {
         "xlrd": _XlrdReader,
         "openpyxl": _OpenpyxlReader,
         "odf": _ODFReader,
         "pyxlsb": _PyxlsbReader,
     }
 
-    def __init__(self, path_or_buffer, engine=None):
+    def __init__(
+        self, path_or_buffer, storage_options: StorageOptions = None, engine=None
```
Comment (TomAugspurger) [inline pandas/io/excel/_base.py:861]: Put storage_options after engine, for users passing that as a position?

File: pandas/tests/io/conftest.py [hunk 7]
```diff
@@ -56,6 +66,9 @@ def s3_resource(tips_file, jsonl_file, feather_file):
         os.environ.setdefault("AWS_SECRET_ACCESS_KEY", "foobar_secret")
 
         moto = pytest.importorskip("moto")
```
Comment (TomAugspurger) [inline pandas/tests/io/conftest.py:68]:
```suggestion
        moto = pytest.importorskip("moto", minversion="1.3.14")
```

File: pandas/tests/io/conftest.py [hunk 8]
```diff
@@ -1,4 +1,8 @@
+from distutils.version import LooseVersion
```
Comment (TomAugspurger) [inline pandas/tests/io/conftest.py:1]:
```suggestion
```

File: pandas/tests/io/conftest.py [hunk 9]
```diff
@@ -56,6 +66,9 @@ def s3_resource(tips_file, jsonl_file, feather_file):
         os.environ.setdefault("AWS_SECRET_ACCESS_KEY", "foobar_secret")
 
         moto = pytest.importorskip("moto")
+        pytest.importorskip("flask")  # server mode needs flask too
+        if LooseVersion(moto.__version__) < LooseVersion("1.3.14"):
+            pytest.skip("Moto too old")
```
Comment (TomAugspurger) [inline pandas/tests/io/conftest.py:71]:
```suggestion
```

File: pandas/tests/io/parser/test_network.py [hunk 10]
```diff
@@ -56,50 +56,62 @@ def tips_df(datapath):
 @td.skip_if_not_us_locale()
 class TestS3:
     @td.skip_if_no("s3fs")
-    def test_parse_public_s3_bucket(self, tips_df):
+    def test_parse_public_s3_bucket(self, tips_df, s3so):
```
Comment (TomAugspurger) [inline pandas/tests/io/parser/test_network.py:59]: Why does this test (and similar ones) need to change? I understand the changes to the moto-based ones, but not these which (I think) aren't using moto.

**Review by TomAugspurger — COMMENTED** [review]
Thanks for working on this. A few questions.

File: pandas/tests/io/conftest.py (hunk 7 above)
Comment (martindurant) [reply pandas/tests/io/conftest.py:68]: I didn't know you could do that, would have saved me keystrokes in a lot of places

File: pandas/tests/io/parser/test_network.py (hunk 10 above)
Comment (martindurant) [reply pandas/tests/io/parser/test_network.py:59]: `@pytest.mark.usefixtures("s3_resource")` a few lines above was installing the mocks in the previous version. I don't know what the intent was, but these tests never touched real S3 (and there is no public "pandas-test" bucket).

Comment (TomAugspurger) [reply pandas/tests/io/parser/test_network.py:59]: Ahh gotcha. Thanks!

File: pandas/io/excel/_base.py (hunk 6 above)
Comment (martindurant) [reply pandas/io/excel/_base.py]: done

File: pandas/tests/io/conftest.py (hunk 7 above)
Comment (martindurant) [reply pandas/tests/io/conftest.py]: done

File: pandas/tests/io/conftest.py (hunk 8 above)
Comment (martindurant) [reply pandas/tests/io/conftest.py:1]: done

File: pandas/tests/io/conftest.py (hunk 9 above)
Comment (martindurant) [reply pandas/tests/io/conftest.py]: done

**Comment by TomAugspurger** [comment]
@martindurant the old pyarrow seems to be coming from defaults. Perhaps only conda-forge has new enough packages for us.


So maybe make sure that all the pyarrows have `>=0.12.0` (or whatever you need). That'll get us what we need as long as `conda-forge` is one of the channels in the env.

**Comment by martindurant** [comment]
Interestingly, travis-37-locale doesn't have pyarrow listed at all (but the newest version on defaults is 0.16.0)

**Review by TomAugspurger — APPROVED** [review]
LGTM, will merge later today.

**Pull request merged**